File: Algorithms_and_data_structure/101/pa2/List.c
// Ruikang Tao
// rtao6
// PA2

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

//structs
typedef struct NodeObj{
	int data;
	struct NodeObj* next;
	struct NodeObj* prev;
} NodeObj;

//Node type
typedef NodeObj* Node;

//ListObj type
typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int length;
	int curIndex;
} ListObj;

// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object.
// Private.
Node newNode(int data){
	Node N = malloc(sizeof(NodeObj));
	N->data = data;
	N->next = NULL;
	N->prev = NULL;
	return(N);
}

// freeNode()
// Frees heap memory pointed to by *fN, sets *fN to NULL.
// Private.
void freeNode(Node* fN){
	if(fN != NULL && *fN != NULL){
		free(*fN);
		*fN = NULL;
	}
}

// newList()
// Returns reference to new empty List object
List newList(void){
	List L;
	L = malloc(sizeof(ListObj));
	L->front = NULL;
	L->back = NULL;
	L->length = 0;
	L->curIndex = -1;
	L->cursor = NULL;
	return L;
}

// freeList()
// Frees all heap memory associated with List *fL, and sets *fL to NULL
void freeList(List* fL){
	if(fL == NULL || *fL == NULL){
		return;
	}
	while(length(*fL) > 0){
		deleteBack(*fL);
	}
	free(*fL);
	*fL = NULL;
}

// Access functions -----------------------------------------------------------

// length()
// Returns the number of elements in this List.
int length(List L){
	if(L == NULL){
		printf("List Error: NULL List reference.\n");
		exit (1);
	}
	return L->length;
}

// getIndex()
// Returns the number of elements in this List.
// otherwise returns -1.
int index(List L){
	if(L == NULL){
		printf("List Error: NULL List reference.\n");
		exit (1);
	}
	return L->curIndex;
}

// front()
// Returns front element. Pre: length()>0
int front(List L){
	if(L == NULL){
		printf("List Error: NULL List reference.\n");
		exit (1);
	}
	if(L->length == 0) {
		printf("List Error: Empty List.\n");
		exit (1);		
	} else{
		return L->front->data;
	}
}

// back()
// Returns back element. Pre: length()>0
int back(List L){
	if(L == NULL){
		printf("List Error: NULL List reference.\n");
		exit (1);
	} else if(L->length == 0){
		printf("List Error: Empty List.\n");
		exit (1);		
	} else{
		return L->back->data;
	}
}

// get()
// Returns cursor element. Pre: length()>0, index()>=0
int get(List L){
	if(L == NULL){
		printf("List Error: NULL List reference.\n");
		exit (1);
	} else if(L->cursor == NULL){
		printf("List Error: Cursor undefined.\n");
		exit (1);		
	} else{
		return (L->cursor->data);
	}
}

// equals()
// Returns true if and only if this List and L are the same
// integer sequence. The states of the cursors in the two Lists
// are not used in determining equality.
int equals(List A, List B){
	int boo = 1;
	if(A == NULL || B == NULL){
		printf("List Error: calling equals() on NULL list reference.\n");
		exit(1);
	}
	if(A->length != B->length){
		return 0;
	}
	Node a = NULL;
	Node b = NULL;
	a = A->front;
	b = B->front;
	while(boo && a != NULL){
		boo = (a->data == b->data);
		a = a->next;
		b = b->next;
	}
	return boo;
}


// Manipulation procedures ----------------------------------------------------

// clear()
// Resets this List to its original empty state.
void clear(List L){
	Node p = L->front;
	while(p != NULL){
		Node l = p;
		p = p->next;
		freeNode(&l);
	}
	L->front = L->back = NULL;
	L->cursor = NULL;
	L->curIndex = -1;
	L->length = 0;
}

// moveFront()
// If List is non-empty, places the cursor under the front element,
// otherwise does nothing.
void moveFront(List L){
	if(L == NULL) {
		printf("List error: calling moveFront() on NULL List reference.\n");
        exit(1);
	}
	if(L->length > 0){
		L->cursor = L->front;
		L->curIndex = 0;
	}
}

// moveBack()
// If List is non-empty, places the cursor under the back element,
// otherwise does nothing.
void moveBack(List L){
	if(L == NULL) {
		printf("List error: calling moveBack() on NULL List reference.\n");
        exit(1);
	}
	if(L->length > 0){
		L->cursor = L->back;
		L->curIndex = L->length - 1;
	}
}

// movePrev()
// If cursor is defined and not at front, moves cursor one step toward
// front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing.
void movePrev(List L){
	if (L == NULL){
		printf("List Error: calling movePrev() on NULL List reference.\n");
		exit (1);
	} else if(L->cursor == L->front || L->curIndex == -1){
		L->cursor = NULL;
		L->curIndex = -1;
	} else{
		L->cursor = L->cursor->prev;
		L->curIndex--;
	}
}

// moveNext()
// If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
void moveNext(List L){
	if (L == NULL){
		printf("List Error: calling moveNext() on NULL List reference.\n");
		exit (1);
	} else if(L->cursor == L->back || L->curIndex == -1){
		L->cursor = NULL;
		L->curIndex = -1;
	} else{
		L->cursor = L->cursor->next;
		L->curIndex++;
	}
}

// prepend()
// Insert new element into this List. If List is non-empty,
// insertion takes place before front element.
void prepend(List L, int data){
	Node n = newNode(data);
	if(L == NULL){
		printf("List Error: calling prepend() on NULL List reference.\n");
		exit (1);
	}
	if(L->length == 0){
		L->front = L->back = n;
	} else{
		L->front->prev = n;
		n->next = L->front;
		L->front = n;
	}
	L->length++;
	if(L->curIndex >= 0){
		L->curIndex++;
	}
}

// append()
// Insert new element into this List. If List is non-empty,
// insertion takes place after back element.
void append(List L, int data){
	Node n = newNode(data);
	if(L == NULL){
		printf("List Error: calling append() on NULL List reference.\n");
		exit (1);
	}
	if(L->length == 0){
		L->front = L->back = n;
	} else{
		L->back->next = n;
		n->prev = L->back;
		L->back = n;
	}
	L->length++;
	/*
	if(L->curIndex >= 0){
		L->curIndex++;
	}
	*/
}

// insertBefore()
// Insert new element before cursor. 
// Pre: length()>0, index()>=0
void insertBefore(List L, int data){
	Node n = newNode(data);
	if(L == NULL){
		printf("List Error: calling insertBefore() on NULL List reference.\n");
		exit (1);
	}
	if(L->length <= 0){
		printf("List Error: calling insertBefore() on an empty list.\n");
		exit (1);
	}
	if(L->cursor == NULL || L->curIndex < 0){
		printf("List Error: calling insertBefore() on undefined cursor.\n");
		exit (1);
	}
	if(L->cursor == L->front || L->curIndex == 0){
		prepend(L,data);
	} 
	else{
		n->next = L->cursor;
		n->prev = L->cursor->prev;
		L->cursor->prev->next = n;
		L->cursor->prev = n;
		L->length++;
		L->curIndex++;
	}
}

// insertAfter()
// Inserts new element after cursor.
// Pre: length()>0, index()>=0
void insertAfter(List L, int data){
	Node n = newNode(data);
	if(L == NULL){
		printf("List Error: calling insertBefore() on NULL List reference.\n");
		exit (1);
	}
	if(L->length <= 0){
		printf("List Error: calling insertBefore() on an empty list.\n");
		exit (1);
	}
	if(L->cursor == NULL || L->curIndex < 0){
		printf("List Error: calling insertBefore() on undefined cursor.\n");
		exit (1);
	}
	if(L->cursor == L->back || L->curIndex == length(L) - 1){
		append(L,data);
	}
	else{
		n->next = L->cursor->next;
		n->prev = L->cursor;
		L->cursor->next->prev = n;
		L->cursor->next = n;
		L->length++;
	}
}

// deleteFront()
// Deletes the front element. Pre: length()>0
void deleteFront(List L){
	if(L == NULL){
		printf("List Error: calling deleteFront() on NULL List reference.\n");
		exit (1);
	}
	if(L->length <= 0){
		printf("List Error: calling deleteFront() on an empty list.\n");
		exit (1);
	}
	Node n = L->front;
	if(L->length == 1){
		L->front = NULL;
	}
	else{
		Node temp = L->front->next;
		temp->prev = NULL;
		L->front = temp;
	}
	
	if(L->curIndex != -1){
		L->curIndex--;
	}
	
	L->length--;
	freeNode(&n);
}

// deleteBack()
// Deletes the back element. Pre: length()>0
void deleteBack(List L){
	if(L == NULL){
		printf("List Error: calling deleteBack() on NULL List reference.\n");
		exit (1);
	}
	if(L->length <= 0){
		printf("List Error: calling deleteBack() on an empty list.\n");
		exit (1);
	}
	Node n = L->back;
	if(L->length == 1){
		L->back = L->front = NULL;
		L->curIndex = -1;
	} else if(L->length >1 && L->curIndex == L->length - 1){
		L->back = L->back->prev;
		L->back->next = NULL;
		L->curIndex = -1;
	} else{
		Node temp = L->back->prev;
		temp->next = NULL;
		L->back = temp;
	}

	/*
	if(L->curIndex != -1){
		L->curIndex--;
	}
	*/
	
	L->length--;
	freeNode(&n);
}

// delete()
// Deletes cursor element, making cursor undefined.
// Pre: length()>0, index()>=0
void delete(List L){
	if(L == NULL){
		printf("List Error: calling delete() on NULL List reference.\n");
		exit (1);
	}
	if(L->length <= 0){
		printf("List Error: calling delete() on an empty list.\n");
		exit (1);
	}
	if (L->curIndex < 0){
		printf("List Error: undefined cursor.\n");
	}
	Node n = L->cursor;
	if (n == L->front){
		deleteFront(L);
	} else if(n == L->back){
		deleteBack(L);
	} else{
		Node before = L->cursor->prev;
		Node after = L->cursor->next;
		before->next = after;
		after->prev = before;
		L->curIndex = -1;
		L->length--;
		freeNode(&n);
	}
	L->cursor = NULL;
	L->curIndex = -1;
}

// printList()
// prints the elements in the list
void printList(FILE* out, List L){
	if(L == NULL){
		printf("List Error: calling printList() on NULL List reference.\n");
		exit (1);
	}
	Node p = NULL;
	for(p = L->front; p != NULL; p = p->next){
		fprintf(out, "%d ", p->data);
	}
}

// copyList()
// copies the element into a new list
List copyList(List L){
	if(L == NULL){
		printf("List Error: calling copyList() on NULL List reference.\n");
		exit (1);
	}
	List xin = newList();
	Node temp = L->front;
	while(temp != NULL){
		append(xin, temp->data);
		temp = temp->next;
	}
	xin->curIndex = -1;
	xin->cursor = NULL;
	L->curIndex = 0;
	return xin;
}

File: Algorithms_and_data_structure/101/pa3/List.java
/**
 *Ruikang Tao
 *rtao6
 *Programming Assignment 3
 */

class List{
	private class Node{
		// inner class fields
		Object data;
		Node prev;
		Node next;
		
		// inner class constructor
		Node(Object data){
			this.data = data;
			prev = null;
			next = null;
		}
		
		// toString()
		public String toString(){
			return String.valueOf(data);
		}
	}

	// private fields
	private Node front;
	private Node back;
	private Node cursor;
	private int length;
	private int cursorIndex;
	
	// Constructor
	public List(){
		front = back = null;
		cursor = null;
		length = 0;
		cursorIndex = -1;
	}
	

	// Access functions:
	
	// length()
	// returns length of List
	int length(){
		return length;
	}
	
	// index()
	// If cursor is defined, returns index of the cursor element
	// otherwise return -1
	int index(){
		return cursorIndex;
	}
	
	// front()
	// Returns front element
	// Pre: length()>0
	Object front(){
		if(length <= 0){
			throw new RuntimeException("front() called on an Empty List.");
		}
		return front.data;
	}
	
	// back()
	// Returns back element 
	// Pre: length()>0
	Object back(){
		if(length <= 0){
			throw new RuntimeException("back() called on an Empty List.");
		}
		return back.data;
	}
	
	// get()
	// Returns cursor element
	// Pre: length()>0, index()>=0
	Object get(){
		if(length <= 0){
			throw new RuntimeException("get() called on an Empty List.");
		}
		if(this.index() <= -1){
			throw new RuntimeException("get() called on undefined cursor index.");
		}
		return cursor.data;
	}
	
	// equals()
	// Returns true if this List and L are the same int seq.
	public boolean equals(Object ha){

		List that = null;
		if(ha instanceof List){
			that = (List)ha;
			if(this.length == that.length){
				Node tmp1 = this.front;
				Node tmp2 = that.front;
				while(tmp1 != null && tmp2 != null){
					if(tmp1.data.equals(tmp2.data)){
						tmp1=tmp1.next;
						tmp2=tmp2.next;
					} else{
						return false;
					}
				}
				return true;
			} else{
				return false;
			}
		} else{
			return false;
		}
	}
	

	// Manipulation procedures:
	
	// clear()
	// Resets this List to its original empty state
	void clear(){
		Node front = null;
		Node back = null;
		cursor = null;
		length = 0;
		cursorIndex = -1;
	}
	
	// moveFront()
	// Places cursor under front element
	// Pre: length != 0
	void moveFront(){
		if(length() > 0){
			cursor = this.front;
			cursorIndex = 0;
		}
	}
	
	// moveBack()
	// Places cursor under back element
	// Pre: length != 0
	void moveBack(){
		if(length() > 0){
			cursor = this.back;
			cursorIndex = length-1;
		}
	}
	
	// movePrev()
	// Moves cursor one step toward front of list
	// cursor must be defined, does nothing otherwise
	// cursor falls off list when called on front
	void movePrev(){
		if(this.index() != -1 && this.index() > 0){
			cursor = cursor.prev;
			cursorIndex--;
		}else if(this.index() != -1 && this.index() == 0){
			cursor = null;
			cursorIndex = -1;
		}
	}
	
	// moveNext()
	// Moves cursor one step toward back of list
	// cursor must be defined, does nothing otherwise 
	// cursor falls of list when called on back
	void moveNext(){
		if(this.index() != -1 && cursorIndex < this.length()-1){
			cursor = cursor.next;
			cursorIndex++;
		}else if(this.index() != -1 && this.index() == this.length()-1){
			cursor = null;
			cursorIndex = -1;
		}
	}
	
	// prepend()
	// Insert new element into list before front element
	void prepend(Object data){
		Node N = new Node(data);
		if(length < 0){
			throw new RuntimeException("prepend() called on an Empty List.");
		}else if(length == 0){
			front = back = N;
			length ++;
		}else{
			front.prev = N;
			N.next = front;
			N.prev = null;
			front = N;
			length ++;
		}
		if(index() != -1){
			cursorIndex ++;
		}
	}
	
	// append()
	// Inserts new element into list after back element
	void append(Object data){
		Node N = new Node(data);
		// if List is empty 
		if(length == 0){
			front = back = N;
		}else{
			// if List non-empty, insert after back element
			N.prev = back;
			N.next = null;
			back.next = N;
			back = N;
		}
		length++;
	}
	
	// insertBefore()
	// Insert new element before cursor
	// Pre: length()>0, index()>=0
	void insertBefore(Object data){

		if (length() <= 0) {
			throw new RuntimeException ("Error: insertBefore() called on an Empty List.");
		} else if (index() < 0) {
			throw new RuntimeException ("Error: Cursor Undefined");
		}
		Node newNode = new Node (data);
		if (index() == 0) {
			prepend(data);
		} else {
			newNode.next = cursor;
			newNode.prev = cursor.prev;
			cursor.prev.next = newNode;
			cursor.prev = newNode;
			cursorIndex++;
			length++;
		}
	}
	
	// insertAfter()
	// Insert new element after cursor
	// Pre: length()>0, index()>=0
	void insertAfter(Object data){

		if (length() <= 0) {
			throw new RuntimeException ("Error: Empty List.");
		} else if (index() < 0) {
			throw new RuntimeException ("Error: Cursor Undefined.");
		}
		Node newNode = new Node (data);
		if (index() == length()-1) {
			append(data);
		} else {
			newNode.next  = cursor.next;
			newNode.prev = cursor;
			cursor.next = newNode;
			cursorIndex++;
			length++;
		}
	}
	
	// deleteFront()
	// Deletes the front element
	// Pre: length()>0
	void deleteFront(){
		if(length <= 0){
			throw new RuntimeException("Error: Empty List.");
		}
		if(this.length > 1){
			front = front.next;
			front.prev = null;
			cursorIndex--;
		}else{
			front = back = null;
			cursorIndex = -1;
		}
		length--;
	}
	
	// deleteBack()
	// Deletes the back element
	// Pre: length()>0
	void deleteBack(){
		if(length <= 0){
			throw new RuntimeException("Error: Empty List.");
		}
		if(this.length > 1 && this.index() != length-1){
			Node temp = back.prev;
			temp.next = null;
			back = temp;
		}else if(this.length() == 1){
			front = back = null;
			cursorIndex = -1;
		}else if(this.length > 1 && this.index() == length-1){
			back = back.prev;
			back.next = null;
			cursorIndex = -1;
		}
		length--;
	}
	
	// delete()
	// Deletes the cursor element, making cursor undef
	// Pre: length()>0, index()>=0
	void delete(){
		if (index() == -1) {
			throw new RuntimeException ("Error: delete() called on list with an undefined cursor.");
		}
		Node temp = cursor;
		if (cursor == back) {
			deleteBack();
		} else if (cursor == front) {
			deleteFront();
		} else{
			cursor.prev.next = cursor.next;
			cursor.next.prev = cursor.prev;
			cursor = null;
			temp = null;
			cursorIndex = -1;
			length--;
		}
	}
	
	
	// Other methods:
	
	// toString()
	// Overrides Object's toString method
	// Returns a String representation of this List
	public String toString(){
		Node N = front;
		String str = "";
		while(N != null){
			str += N.toString();
			N = N.next;
		}
		return str;
	}
}